### PR TITLE
#3414 Credit Limit check in Sales Order 

### PR DESCRIPTION
**Pull request by cristinamghita**
#3414 Credit Limit check in Sales Order

### CODE REVIEW

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java [hunk 1]
```diff
@@ -60,4 +70,28 @@ public void createBPartnerStatsRecord(@NonNull final I_C_BPartner bpartner)
...
 		Services.get(IBPartnerStatsDAO.class).retrieveBPartnerStats(bpartner);
 	}
+
+	@ModelChange(timings = { ModelValidator.TYPE_BEFORE_CHANGE}, ifColumnsChanged = {I_C_BPartner.COLUMNNAME_C_BP_Group_ID})
+	public void updateBPartnerStatsRecord(@NonNull final I_C_BPartner bpartner)
+	{
+		final IBPartnerStats bpartnerStats = Services.get(IBPartnerStatsDAO.class).retrieveBPartnerStats(bpartner);
+		final I_C_BP_Group bpGroup = bpartner.getC_BP_Group();
+		final String creditStatus = bpGroup.getSOCreditStatus();
+		if (!Check.isEmpty(creditStatus,true))
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: don't we have to update BP's stats even when creditStatus is empty?

Comment (teosarca) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: also why not moving this code to BPartnerStatsBL?

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java [hunk 2]
```diff
@@ -60,4 +70,28 @@ public void createBPartnerStatsRecord(@NonNull final I_C_BPartner bpartner)
...
+			setSO_CreditStatus(bpartnerStats, creditStatus);
+
+			// make sure the status is correct
+			Services.get(IBPartnerStatisticsUpdater.class)
+			.updateBPartnerStatistics(Env.getCtx(), Collections.singleton(bpartner.getC_BPartner_ID()), ITrx.TRXNAME_None);
+
+		}
+	}
+
+	private void setSO_CreditStatus(@NonNull final IBPartnerStats bpartnerStats, @NonNull final String creditStatus)
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: i don't see the benefit of this method.
imho it's introducing more confusion than making things more clear. When you read the name you don't know what it does... does it also save the record or only it sets that field?!?

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 3]
```diff
@@ -41,26 +51,72 @@
 @Repository
 public class BPartnerCreditLimitRepository
 {
-	public BigDecimal retrieveCreditLimitByBPartnerId(final int bpartnerId, final Timestamp date)
+	public BigDecimal getCreditLimitByBPartner(@NonNull final I_C_BPartner bpartner, @NonNull final Timestamp date)
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: why not using bpartnerId?

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 4]
```diff
@@ -41,26 +51,72 @@
...
+		}
 
-		final I_C_BPartner_CreditLimit bpCreditLimit = Services.get(IQueryBL.class)
-				.createQueryBuilder(I_C_BPartner_CreditLimit.class)
-				.addEqualsFilter(I_C_BPartner_CreditLimit.COLUMNNAME_C_BPartner_ID, bpartnerId)
+		sort(bpCreditLimits);
+		return bpCreditLimits.get(0).getAmount();
+	}
+
+	public List<I_C_BPartner_CreditLimit> retrieveCreditLimitsByBPartner(@NonNull final I_C_BPartner bpartner, @NonNull final Timestamp date)
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: why not using bpartnerId?

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 5]
```diff
@@ -41,26 +51,72 @@
...
+		final Supplier<I_C_CreditLimit_Type> recordLoader = () -> retrieveRecordFromDB(C_CreditLimit_Type_ID);
+		final I_C_CreditLimit_Type type = cache_creditLimitById.get(C_CreditLimit_Type_ID, recordLoader);
+
+		return CreditLimitTypePOJO.builder()
+				.C_CreditLimit_Type_ID(type.getC_CreditLimit_Type_ID())
+				.seqNo(type.getSeqNo())
+				.build();
+	}
+
+	private I_C_CreditLimit_Type retrieveRecordFromDB(final int C_CreditLimit_Type_ID)
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: here u shall return the POJO...

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 6]
```diff
@@ -41,26 +51,72 @@
...
+	private static class CreditLimitTypePOJO
+	{
+		private final int seqNo;
+		private final int C_CreditLimit_Type_ID;
+	}
+
+	private CreditLimitTypePOJO getCreditLimitTypeById(final int C_CreditLimit_Type_ID)
+	{
+		final Supplier<I_C_CreditLimit_Type> recordLoader = () -> retrieveRecordFromDB(C_CreditLimit_Type_ID);
+		final I_C_CreditLimit_Type type = cache_creditLimitById.get(C_CreditLimit_Type_ID, recordLoader);
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: pls merge these 2 lines... and use CCache.getOrLoad

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 7]
```diff
@@ -41,26 +51,72 @@
...
+		final Comparator<I_C_BPartner_CreditLimit> ORDERING_BPCreditLimitByTypeSeqNoReversed = ORDERING_BPCreditLimitByTypeSeqNo.reversed();
+
+		final Comparator<I_C_BPartner_CreditLimit> ORDERING_BPCreditLimitByDateFrom = Comparator.comparing(bpCreditLimit -> bpCreditLimit.getDateFrom());
+		final Comparator<I_C_BPartner_CreditLimit> ORDERING_BPCreditLimitByDateFromReversed = ORDERING_BPCreditLimitByDateFrom.reversed();
 
-		return bpCreditLimit.getAmount();
+		bpCreditLimits.sort(ORDERING_BPCreditLimitByTypeSeqNoReversed.thenComparing(ORDERING_BPCreditLimitByDateFromReversed));
+	}
+
+	private final CCache<Integer, I_C_CreditLimit_Type> cache_creditLimitById = CCache.newCache(I_C_CreditLimit_Type.Table_Name + "#CreditLimitType#by#Id", 10, CCache.EXPIREMINUTES_Never);
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: move this on top... fields shall be on top...

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java [hunk 8]
```diff
@@ -41,26 +51,72 @@
...
+		final List<I_C_BPartner_CreditLimit> bpCreditLimits = retrieveCreditLimitsByBPartner(bpartner, date);
+		if (bpCreditLimits.isEmpty())
+		{
+			return BigDecimal.ZERO;
+		}
 
-		final I_C_BPartner_CreditLimit bpCreditLimit = Services.get(IQueryBL.class)
-				.createQueryBuilder(I_C_BPartner_CreditLimit.class)
-				.addEqualsFilter(I_C_BPartner_CreditLimit.COLUMNNAME_C_BPartner_ID, bpartnerId)
+		sort(bpCreditLimits);
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/BPartnerCreditLimitRepository.java]: not OK, you have no guarantees that the List is mutable.
pls write this method using streams....

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java [hunk 9]
```diff
@@ -60,4 +70,28 @@ public void createBPartnerStatsRecord(@NonNull final I_C_BPartner bpartner)
...
+
+	@ModelChange(timings = { ModelValidator.TYPE_BEFORE_CHANGE}, ifColumnsChanged = {I_C_BPartner.COLUMNNAME_C_BP_Group_ID})
+	public void updateBPartnerStatsRecord(@NonNull final I_C_BPartner bpartner)
+	{
+		final IBPartnerStats bpartnerStats = Services.get(IBPartnerStatsDAO.class).retrieveBPartnerStats(bpartner);
+		final I_C_BP_Group bpGroup = bpartner.getC_BP_Group();
+		final String creditStatus = bpGroup.getSOCreditStatus();
+		if (!Check.isEmpty(creditStatus,true))
+		{
+			setSO_CreditStatus(bpartnerStats, creditStatus);
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: also i am a bit puzzled... wondering, isn't the job of the updater to set the SO_CreditStatus, and mainly, to update the C_BPartner_Stats?

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/IBPartnerStats.java [hunk 10]
```diff
@@ -62,4 +62,6 @@
 	 */
 	int getC_BPartner_ID();
 
+	I_C_BPartner_Stats getC_BPartner_Stats();
```
Comment (teosarca) [inline de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/service/IBPartnerStats.java]: hmmm... wondering what's the role of IBPartnerStats if we have this getter??!!

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/spi/AbstractInvoiceCandidateHandler.java [hunk 11]
```diff
@@ -75,7 +76,22 @@ public Object getModelForInvoiceCandidateGenerateScheduling(final Object model)
...
+
+	@Override
+	public void setLineNetAmt(@NonNull final I_C_Invoice_Candidate ic)
+	{
+		final boolean force = ic.getQtyToInvoice().signum() ==  0 ? true : false;
+		final BigDecimal netAmt = computeNetAmt(ic, force);
+		ic.setLineNetAmt(netAmt);
+	}
+
+	private BigDecimal computeNetAmt(@NonNull final I_C_Invoice_Candidate ic, final boolean force)
```
Comment (teosarca) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/spi/AbstractInvoiceCandidateHandler.java]: pls document what `force` means.
also pls try to find a better name.

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java (hunk 1 above)
Comment (cristinamghita) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: The credit status is updated when changes are made regarding partner credit limit or document.
If the credit status in bpartner group is empty, I don;t want to touch partner's credit status.

Comment (cristinamghita) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: but, yes, the logic we could have it in BPartnerStatsBL. Better place there

File: de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java (hunk 9 above)
Comment (cristinamghita) [reply de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/bpartner/model/interceptor/C_BPartner.java]: yes; But, at first, when we create the partner, we need the creditstatus based on bpartner group.

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandidateHandlerBL.java [hunk 12]
```diff
@@ -135,14 +135,7 @@ public IInvoiceCandidateHandler mkInstance(final I_C_ILCandHandler handlerRecord
...
 	{
-		Services.get(ITrxManager.class).run(new TrxRunnable()
-		{
-			@Override
-			public void run(final String trxName) throws Exception
-			{
-				createInvoiceCandidates(handlerRecords, InvoiceCandidateHandlerBL.NO_MODEL);
-			}
-		});
+		Services.get(ITrxManager.class).run((TrxRunnable)trxName -> createInvoiceCandidates(handlerRecords, InvoiceCandidateHandlerBL.NO_MODEL));
```
[the remark this answers is not in the document]
Comment (cristinamghita) [reply de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandidateHandlerBL.java]: eclipse is doing automatically; no I don;t need the cast

**Pull request merged**